### PR TITLE
create a factory for fixed sparstity euclidean matrices

### PR DESCRIPTION
**Pull request by bamdevm**
Euclidean matrices with fixed sparsity patttern.

### CODE REVIEW

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+% Returns M, a structure describing the Euclidean space of real matrices,
+% equipped with the standard Frobenius distance and associated trace inner
+% product, as a manifold for Manopt.
+%
+% See also: euclideanfactory euclideancomplexfactory
+
+% This file is part of Manopt: www.manopt.org.
+% Original author: Bamdev Mishra, Mar. 28, 2019.
+% Change log: 
+%
```
Comment (NicolasBoumal) [inline tests/bmsparseeuclidean/sparseeuclideanfactory.m]:
```suggestion
%    May 3, 2019 (NB): adapted many functions to take better advantage of sparsity.
```

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 2]
```diff
@@ -0,0 +1,80 @@
+function M = sparseeuclideanfactory(A)
+% Returns a manifold struct to optimize over real matrices with sparsity pattern that of A.
+%
+% function M = sparseeuclideanfactory(A)
+%
+% Returns M, a structure describing the Euclidean space of real matrices,
+% equipped with the standard Frobenius distance and associated trace inner
+% product, as a manifold for Manopt.
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: Done

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+%
+% See also: euclideanfactory euclideancomplexfactory
+
+% This file is part of Manopt: www.manopt.org.
+% Original author: Bamdev Mishra, Mar. 28, 2019.
+% Change log: 
+%
+	
+	dimensions_vec = size(A);
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: Done.

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 4]
```diff
@@ -0,0 +1,80 @@
+function M = sparseeuclideanfactory(A)
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: Done

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	S = sparse(I, J, ones(length(I),1), dimensions_vec(1), dimensions_vec(2));
+	nvals = length(I);
+      
+    M.size = @() dimensions_vec;
+    
+    M.name = @() sprintf('Euclidean space R^(%s) with fixed sparsity pattern containg %s non-zero entries', num2str(dimensions_vec), num2str(nvals));
+    
+    M.dim = @() nvals;
+    
+    M.inner = @(x, d1, d2) d1(:).'*d2(:);
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: Will dot(d1(:), d2(:)) work better?

Ideally, as you suggested, we should only store indices. 

But, we can stick to this d1(:).'*d2(:) for the time being unless we have better.

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    
+    M.norm = @(x, d) norm(d(:), 'fro');
+    
+    M.dist = @(x, y) norm(x(:) - y(:), 'fro');
+    
+    M.typicaldist = @() sqrt(prod(dimensions_vec));
+    
+    M.proj = @(x, d) S.*d;
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: Yes.

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        else
+            y = x + d;
+        end
+    end
+    
+    M.retr = M.exp;
+	
+	M.log = @(x, y) y-x;
+
+    M.hash = @(x) ['z' hashmd5(x(:))];
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: thats nice. I missed it completely.

File: tests/bmsparseeuclidean/sparseeuclideanfactory.m [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    
+    M.zerovec = @(x) zeros(dimensions_vec);
+    
+    M.transp = @(x1, x2, d) d;
+    M.isotransp = M.transp; % the transport is isometric
+    
+    M.pairmean = @(x1, x2) .5*(x1+x2);
+    
+    M.vec = @(x, u_mat) u_mat(:);
+    M.mat = @(x, u_vec) reshape(u_vec, dimensions_vec);
```
[the remark this answers is not in the document]
Comment (bamdevm) [reply tests/bmsparseeuclidean/sparseeuclideanfactory.m]: k should be nvals, right?

**Comment by bamdevm** [comment]
Hello Nicolas,

Thanks.

I guess, we still need to copy the factory file from the tests to the
manifolds folder.

Regards,
Bamdev

**Comment by NicolasBoumal** [comment]
Done!

**Pull request merged**